Upstream fpr

nmanthey opened the pull request:
The first commit of this series addresses https://github.com/diffblue/cbmc/issues/1783
The second one should allow to make function pointer removal somewhat faster.

nmanthey committed:
called_functions: use unordered_set

Instead of using an ordered set, use a faster unordered set. This way,
running the algorithms against larger code bases should result in
decreased run times, while on small programs the run time might slightly
increase.

As this function is used in function pointer removal, and in code that
computes unreachable functions, the type change had to be propagated to
the related classes as well.

tautschnig commented on src/goto-programs/compute_called_functions.cpp:
```diff
@@ -91,11 +90,10 @@ std::set<irep_idt> compute_called_functions(
     working_queue.erase(working_queue.begin());
 
-    if(done.find(id)!=done.end())
+    if(functions.find(id)!=functions.end())
```
A few nit picks (placing them here for github doesn't feature commenting on commit messages):
1) Right here: clang-format preferred spacing: add spaces around `!=`.
2) You might actually do:
```
if(!functions.insert(id).second)
  continue;
```
and thereby further improve performance by not duplicating the log-time search of the separate find+insert steps.
2) Commit message 1: Please add "Fixes: #1783" to auto-close the issue.
3) Commit message 2: Please spell out "Function-pointer" (instead of fp) to avoid confusion, e.g., with "floating-point."

tautschnig commented on src/goto-analyzer/unreachable_instructions.cpp:
```diff
@@ -289,9 +289,10 @@ static void xml_output_function(
 static void list_functions(
   const goto_modelt &goto_model,
-  const std::set<irep_idt> called,
+  const T called,
```
I'm aware this will cause more changes, but I'd insist that the template is unnecessary here and everyone should be using the updated interface.

tautschnig commented on src/goto-analyzer/unreachable_instructions.cpp:
```diff
@@ -167,7 +167,7 @@ void unreachable_instructions(
   json_arrayt json_result;
 
-  std::set<irep_idt> called=
+  std::unordered_set<irep_idt> called=
```
Nit pick: introduce a space before `=` to converge towards clang-format's suggested style.

tautschnig commented on src/goto-programs/link_to_library.cpp:
```diff
@@ -45,7 +45,7 @@ void link_to_library(
   while(true)
   {
-    std::set<irep_idt> called_functions=
+    std::unordered_set<irep_idt> called_functions=
```
Nit pick as above: introduce a space before `=`

hannes-steffenhagen-diffblue commented:
Please use `git-clang-format-3.8 HEAD~2` to apply clang-format to these changes.

kroening commented:
Looks good; the template is indeed not necessary, as changing set -> unordered_set isn't a massive burden on users of this API.

nmanthey committed:
function-pointer-removal: drop unused set

The compute_called_functions method maintains two sets, 'done' and
'functions'. The set 'done' is only used to determine whether an element
can be skipped. Since the content of the two sets is the same, the set
'done' can be dropped, and 'functions' can be used instead. Hence, this
commit drops the set 'done'.

Furthermore, the return value of insert includes a bool that indicates
the success of the insert operation. In case insertion failed, the element
has been present already, so that we can merge the find+insert into a
single insert operation.

Fixes: #1783

nmanthey commented:
I updated and rebased.

I fixed the raised concerns, and formatted the code with clang-format. I furthermore removed the template and propagated the unordered_set change further.
